osutil/synctree.go

```diff
@@ -0,0 +1,138 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"os"
+	"path/filepath"
+	"sort"
+	"syscall"
+)
+
+func appendWithPrefix(paths []string, prefix string, filenames []string) []string {
+	for _, filename := range filenames {
+		paths = append(paths, filepath.Join(prefix, filename))
+	}
+	return paths
+}
+
+func removeEmptyDirs(baseDir, relPath string) error {
+	for relPath != "." {
```

Comment:
Why the relative path cannot be the current directory?

Reply:
It seemed appropriate for the icon theme case.  Just as we don't delete `/var/lib/snapd/desktop/applications` when uninstalling a snap, it seemed sensible to do the the same with the base icon directory: treat that as being owned by the snapd package.

And we definitely want to have a stop condition: we don't want to remove directories outside of `baseDir`